lib/class-wp-rest-post-search-handler.php

```diff
@@ -0,0 +1,191 @@
+<?php
+/**
+ * REST API: WP_REST_Post_Search_Handler class
+ *
+ * @package gutenberg
+ * @since 3.3.0
+ */
+
+/**
+ * Core class representing a search handler for posts in the REST API.
+ *
+ * @since 3.3.0
+ */
+class WP_REST_Post_Search_Handler extends WP_REST_Object_Search_Handler {
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 3.3.0
+	 */
+	public function __construct() {
+		$this->type     = 'post';
+		$this->subtypes = array_values( get_post_types( array(
+			'public'       => true,
+			'show_in_rest' => true,
+		), 'names' ) );
+	}
+
+	/**
+	 * Searches the object type content for a given search request.
+	 *
+	 * @since 3.3.0
+	 *
+	 * @param WP_REST_Request $request Full REST request.
+	 * @return array Associative array containing an `WP_REST_Object_Search_Handler::RESULT_IDS` containing
+	 *               an array of found IDs and `WP_REST_Object_Search_Handler::RESULT_TOTAL` containing the
+	 *               total count for the matching search results.
+	 */
+	public function search_items( WP_REST_Request $request ) {
+
+		// Get the post types to search for the current request.
+		$post_types = $request[ WP_REST_Search_Controller::PROP_SUBTYPE ];
+		if ( in_array( WP_REST_Search_Controller::TYPE_ANY, $post_types, true ) ) {
+			$post_types = $this->subtypes;
+		}
+
+		// Get the public post statuses as only those should be searched.
+		$post_statuses = array_values( get_post_stati( array(
```

Comment:
For parity with core, we should simply use `post_status=>publish`. We can discuss custom statuses at greater depth with #3144

Reply:
I'm still not fully convinced why as the REST API handles post statuses like that in core already, but I'm okay simplifying it.

Reply:
> I'm still not fully convinced why as the REST API handles post statuses like that in core already, but I'm okay simplifying it.

Oh. If that's the case, I'm fine with it as-is.

Reply:
For now, I decided to follow your suggestion and only use `publish` for now. The only problem that brings is that the controller won't be able to search attachments at this point because they typically use `inherit`. Is it acceptable not having that in the first iteration? Pinging @pento for an opinion on this too.

Reply:
Because searching for attachments will always return an empty array for now (which is unexpected), I explicitly removed support for `type=>post` and `subtype=attachment`. Once we add support, we can allow that again.